Docs for roles/role_cfg (#4221)
tarantool/doc: doc/how-to/app/app_roles.rst

.. _application_roles:

Application roles
=================

An application role is a Lua module that implements specific functions or logic.
You can enable or turn off a particular role for certain instances in a :ref:`configuration <configuration>` without restarting these instances.
A role is run when a configuration is loaded or reloaded.

Roles can be divided into the following groups:

> -   Roles provided by Tarantool.

Reviewer: ```suggestion
-   Tarantool's built-in roles.
```
Nit: this sounds more natural to me (if this wording is correct in our terminology).

    For example, the ``config.storage`` role can be used to make a Tarantool replica set act as a :ref:`configuration storage <centralized_configuration_storage_set_up_tarantool>`.
-   Roles provided by third-party Lua modules.
    For example, the `CRUD <https://github.com/tarantool/crud>`__ module provides the ``roles.crud-storage`` and ``roles.crud-router`` roles that enable CRUD operations in a sharded cluster.
-   Custom roles that are developed as a part of a cluster application.
    For example, you can create a custom role to define a stored procedure or implement a supplementary service, such as an email notifier or a replicator.

This section describes how to develop custom roles.
To learn how to enable and configure roles, see :ref:`configuration_application_roles`.

..  NOTE::

    Don't confuse application roles with other role types:

>     -   A role is a container for privileges that can be granted to users. Learn more in :ref:`access_control_concepts_roles`.
>     -   A role of a replica set in regard to sharding. Learn more in :ref:`vshard_config_sharding_roles`.

Reviewer: ```suggestion
    -   A user role is a container for privileges that can be granted to users. Learn more in :ref:`access_control_concepts_roles`.
    -   A sharding role of a replica set in regard to sharding. Learn more in :ref:`vshard_config_sharding_roles`.
```
Just an idea, to decrease readers' confusion a bit

Author: > A user role

Sounds better but a role can also has a role :) So, it's better to keep it as is.

> A sharding role of a replica set in regard to sharding.

`sharding` meets two times in this case and add some confusion.

.. _roles_create_custom_role_config:

Providing a role configuration
------------------------------

A custom role can be configured in the same way as roles provided by Tarantool or third-party Lua modules.
You can learn more from :ref:`configuration_application_roles`.

This example shows how to enable and configure the ``greeter`` role, which is implemented in the next section:

..  literalinclude:: /code_snippets/snippets/config/instances.enabled/application_role_cfg/config.yaml
    :language: yaml
    :start-at: instance001
    :dedent:

The role's configuration provided in ``roles_cfg`` can be accessed when :ref:`validating <roles_create_custom_role_validate>` and :ref:`applying <roles_create_custom_role_apply>` this configuration.

> Given that a role is a :ref:`Lua module <app_server-modules>`, a role's name is passed to ``require()`` to obtain the module.

Reviewer: I'm confused with passive voice here: does it explain what Tarantool does under the hood or instructs me to pass the role's name when I write the app code?

Author: explains what Tarantool does under the hood. Maybe, this should be a note...

When :ref:`developing an application <admin-instance_config-develop-app>`, you can place a file with a role's code next to a cluster's configuration file.

.. _roles_create_custom_role:

Creating a custom role
----------------------

.. _roles_create_custom_role_overview:

Overview
~~~~~~~~

Creating a custom role includes the following steps:

> 1.  Define a function that validates a role's configuration.

Reviewer: I'd provide the function names explicitly in this list.

Author: This list give a general idea about role's functions. The code snippet after this lists the exact names.

2.  Define a function that applies a validated configuration.
3.  Define a function that stops a role.
4.  (Optional) Define roles from which this custom role depends on.

As a result, a role's module should return an object that has corresponding functions and fields specified:

..  code-block:: lua

    return {
        validate = function() -- ... -- end,
        apply = function() -- ... -- end,
        stop = function() -- ... -- end,
        dependencies = { -- ... -- },
    }

The examples below show how to do this.

..  NOTE::

    Code snippets shown in this section are included from the following application: `application_role_cfg <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/config/instances.enabled/application_role_cfg>`_.

.. _roles_create_custom_role_validate:

Validating a role configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To validate a role's configuration, you need to define the :ref:`validate([cfg]) <roles_api_reference_validate>` function.

> The ``cfg`` argument allows you to get access to the provided :ref:`role's configuration <roles_create_custom_role_config>` and check its validity.

Reviewer: ```suggestion
The ``cfg`` argument provides access to the :ref:`role's configuration <roles_create_custom_role_config>` and check its validity.
```
Seems shorter and easier to read

..  literalinclude:: /code_snippets/snippets/config/instances.enabled/application_role_cfg/byeer.lua
    :language: lua
    :dedent:

A role cannot be started without its dependencies.
This means that all the dependencies of a role should be defined in the ``roles`` configuration parameter:

..  code-block:: yaml

    instance001:
      roles: [ greeter, byeer ]

You can find the full example here: `application_role_cfg <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/config/instances.enabled/application_role_cfg>`_.

.. _roles_create_custom_role_init:

Adding initialization code
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can add initialization code to a role by defining and calling a function with an arbitrary name at the top level of a module, for example:

..  code-block:: lua

    local function init()
        -- ... --
    end

    init()

For example, you can :ref:`create spaces <box_schema-space_create>`, define :ref:`indexes <concepts-data_model_indexes>`, or :ref:`grant privileges <configuration_credentials>` to specific users or roles.

See also: :ref:`roles_create_space`.

.. _roles_create_space:

> Specifics of creating spaces

Reviewer: ```suggestion
Creating spaces in roles
```

Author: This applies to creating spaces for applications in general. So, I'm going to move this section outside this topic at some moment.

    you also need to execute schema initialization code from :ref:`apply() <roles_create_custom_role_apply>`.
    To make sure a space is created only once, use the :ref:`if_not_exists <space_opts_if_not_exists>` option.

.. _roles_life_cycle:

Roles life cycle
----------------

A role’s life cycle includes the stages described below.

..  _roles_life_cycle_loading_roles:

1)  *Loading roles*

    On each run, all roles are loaded in the order they are specified in the :ref:`configuration <roles_create_custom_role_config>`.
    This stage takes effect when a role is enabled or an instance with this role is restarted.
    At this stage, a role executes the :ref:`initialization code <roles_create_custom_role_init>`.

    A role cannot be started if it has :ref:`dependencies <roles_create_custom_role_dependencies>` that are not specified in a configuration.

    ..  NOTE::

        Dependencies do not affect the order in which roles are loaded.
        However, the ``validate()``, ``apply()``, and ``stop()`` functions are executed taking dependencies into account.
        Learn more in :ref:`roles_life_cycle_dependencies_specifics`.

..  _roles_life_cycle_stopping_roles:

> 2)  *Stopping roles*

Reviewer: Maybe `Spotting removed roles` or smth like that? the current wording surprised me and took some time to understand which roles we're talking about

.. _roles_life_cycle_applying_role_config:

4)  *Applying a role's configurations*

    At this stage, a configuration for each role is applied using the corresponding :ref:`apply() <roles_api_reference_apply>` function in the same order in which they are specified in the configuration.

All role's functions report an unrecoverable error by throwing an error object.
If an error is thrown in any phase, applying a configuration is stopped.
If starting or stopping a role throws an error, no roles are stopped or started afterward.
An error is caught and shown in :ref:`config:info() <config_api_reference_info>` in the ``alerts`` section.

.. _roles_life_cycle_dependencies_specifics:

Executing functions for dependent roles
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For roles that :ref:`depend <roles_create_custom_role_dependencies>` on each other, their ``validate()``,  ``apply()``, and ``stop()`` functions are executed taking into account the dependencies.
Suppose, there are two independent roles (``role1``, ``role2``) and three roles that depend on each other as follows:

> ..  code-block:: none

Reviewer: Maybe add a verbal explanation? It's unclear to me now.

    ``apply()`` should throw an error if the specified configuration can't be applied.

    ..  NOTE::

        Note that ``apply()`` is not invoked if an instance switches to read-write mode when :ref:`replication.failover <configuration_reference_replication_failover>` is set to ``election`` or ``supervised``.
        You can check an instance state by subscribing to the ``box.status`` event using :ref:`box.watch() <box-watch>`.

    :param cfg: a role's role configuration to be applied.
                This parameter provides access to configuration options defined in :ref:`roles_cfg.\<role_name\> <configuration_reference_roles_cfg>`.
                To get values of configuration options placed outside ``roles_cfg.<role_name>``, use :ref:`config:get() <config_api_reference_get>`.

    See also: :ref:`roles_create_custom_role_apply`

.. _roles_api_reference_stop:

..  function:: stop()

    Stop a role.
    This function is called on configuration reload if the role is removed from ``roles`` for the given instance.

    See also: :ref:`roles_create_custom_role_stop`

.. _roles_api_reference_dependencies:

..  data:: dependencies

>     (Optional) Define a role's dependencies.

Reviewer: ```suggestion
    (Optional) Role's dependencies.
```
Nit: IMO, verbs are not needed in option/parameter/variable descriptions in references.
